// @0x/sol-compiler/lib/src/utils/fs_wrapper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@0x/utils");
const fs = require("fs");
const mkdirp = require("mkdirp");
exports.fsWrapper = {
    readdirAsync: utils_1.promisify(fs.readdir),
    readFileAsync: utils_1.promisify(fs.readFile),
    writeFileAsync: utils_1.promisify(fs.writeFile),
    mkdirpAsync: utils_1.promisify(mkdirp),
    doesPathExistSync: fs.existsSync,
    rmdirSync: fs.rmdirSync,
    removeFileAsync: utils_1.promisify(fs.unlink),
    statAsync: utils_1.promisify(fs.stat),
    appendFileAsync: utils_1.promisify(fs.appendFile),
    accessAsync: utils_1.promisify(fs.access),
    doesFileExistAsync: (filePath) => __awaiter(this, void 0, void 0, function* () {
        try {
            yield exports.fsWrapper.accessAsync(filePath, 
            // node says we need to use bitwise, but tslint says no:
            fs.constants.F_OK | fs.constants.R_OK);
        }
        catch (err) {
            return false;
        }
        return true;
    }),
};
