// @chainlink/contracts/ethers/v0.6/AggregatorValidatorInterfaceFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "ethers/providers";

import { AggregatorValidatorInterface } from "./AggregatorValidatorInterface";

export class AggregatorValidatorInterfaceFactory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): AggregatorValidatorInterface {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as AggregatorValidatorInterface;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "previousRoundId",
        type: "uint256",
      },
      {
        internalType: "int256",
        name: "previousAnswer",
        type: "int256",
      },
      {
        internalType: "uint256",
        name: "currentRoundId",
        type: "uint256",
      },
      {
        internalType: "int256",
        name: "currentAnswer",
        type: "int256",
      },
    ],
    name: "validate",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
];
