// @chainlink/contracts/ethers/v0.6/AggregatorValidatorInterfaceFactory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
class AggregatorValidatorInterfaceFactory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.AggregatorValidatorInterfaceFactory = AggregatorValidatorInterfaceFactory;
const _abi = [
    {
        inputs: [
            {
                internalType: "uint256",
                name: "previousRoundId",
                type: "uint256",
            },
            {
                internalType: "int256",
                name: "previousAnswer",
                type: "int256",
            },
            {
                internalType: "uint256",
                name: "currentRoundId",
                type: "uint256",
            },
            {
                internalType: "int256",
                name: "currentAnswer",
                type: "int256",
            },
        ],
        name: "validate",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
