// @chainlink/contracts/ethers/v0.6/LinkTokenReceiverFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "ethers/providers";

import { LinkTokenReceiver } from "./LinkTokenReceiver";

export class LinkTokenReceiverFactory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): LinkTokenReceiver {
    return new Contract(address, _abi, signerOrProvider) as LinkTokenReceiver;
  }
}

const _abi = [
  {
    inputs: [],
    name: "getChainlinkToken",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "_sender",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "_amount",
        type: "uint256",
      },
      {
        internalType: "bytes",
        name: "_data",
        type: "bytes",
      },
    ],
    name: "onTokenTransfer",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];
