// @chainlink/contracts/ethers/v0.5/AggregatorV3InterfaceFactory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
class AggregatorV3InterfaceFactory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.AggregatorV3InterfaceFactory = AggregatorV3InterfaceFactory;
const _abi = [
    {
        constant: true,
        inputs: [],
        name: "decimals",
        outputs: [
            {
                name: "",
                type: "uint8",
            },
        ],
        payable: false,
        stateMutability: "view",
        type: "function",
    },
    {
        constant: true,
        inputs: [],
        name: "version",
        outputs: [
            {
                name: "",
                type: "uint256",
            },
        ],
        payable: false,
        stateMutability: "view",
        type: "function",
    },
    {
        constant: true,
        inputs: [],
        name: "description",
        outputs: [
            {
                name: "",
                type: "string",
            },
        ],
        payable: false,
        stateMutability: "view",
        type: "function",
    },
    {
        constant: true,
        inputs: [
            {
                name: "_roundId",
                type: "uint80",
            },
        ],
        name: "getRoundData",
        outputs: [
            {
                name: "roundId",
                type: "uint80",
            },
            {
                name: "answer",
                type: "int256",
            },
            {
                name: "startedAt",
                type: "uint256",
            },
            {
                name: "updatedAt",
                type: "uint256",
            },
            {
                name: "answeredInRound",
                type: "uint80",
            },
        ],
        payable: false,
        stateMutability: "view",
        type: "function",
    },
    {
        constant: true,
        inputs: [],
        name: "latestRoundData",
        outputs: [
            {
                name: "roundId",
                type: "uint80",
            },
            {
                name: "answer",
                type: "int256",
            },
            {
                name: "startedAt",
                type: "uint256",
            },
            {
                name: "updatedAt",
                type: "uint256",
            },
            {
                name: "answeredInRound",
                type: "uint80",
            },
        ],
        payable: false,
        stateMutability: "view",
        type: "function",
    },
];
