// @chainlink/contracts/ethers/v0.5/MaliciousChainlinkClientFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, ContractFactory, Signer } from "ethers";
import { Provider } from "ethers/providers";
import { UnsignedTransaction } from "ethers/utils/transaction";

import { TransactionOverrides } from ".";
import { MaliciousChainlinkClient } from "./MaliciousChainlinkClient";

export class MaliciousChainlinkClientFactory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: TransactionOverrides): Promise<MaliciousChainlinkClient> {
    return super.deploy(overrides) as Promise<MaliciousChainlinkClient>;
  }
  getDeployTransaction(overrides?: TransactionOverrides): UnsignedTransaction {
    return super.getDeployTransaction(overrides);
  }
  attach(address: string): MaliciousChainlinkClient {
    return super.attach(address) as MaliciousChainlinkClient;
  }
  connect(signer: Signer): MaliciousChainlinkClientFactory {
    return super.connect(signer) as MaliciousChainlinkClientFactory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): MaliciousChainlinkClient {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as MaliciousChainlinkClient;
  }
}

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32",
      },
    ],
    name: "ChainlinkRequested",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32",
      },
    ],
    name: "ChainlinkFulfilled",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32",
      },
    ],
    name: "ChainlinkCancelled",
    type: "event",
  },
];

const _bytecode =
  "0x608060405260016004556001600655348015601957600080fd5b5060358060276000396000f3fe6080604052600080fdfea165627a7a72305820e5d5b9650b4d8036225bdfc636ecde0b4f8ac6254032925e0f0579885e1ad8dc0029";
