// @chainlink/contracts/ethers/v0.5/OracleInterfaceFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "ethers/providers";

import { OracleInterface } from "./OracleInterface";

export class OracleInterfaceFactory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): OracleInterface {
    return new Contract(address, _abi, signerOrProvider) as OracleInterface;
  }
}

const _abi = [
  {
    constant: false,
    inputs: [
      {
        name: "requestId",
        type: "bytes32",
      },
      {
        name: "payment",
        type: "uint256",
      },
      {
        name: "callbackAddress",
        type: "address",
      },
      {
        name: "callbackFunctionId",
        type: "bytes4",
      },
      {
        name: "expiration",
        type: "uint256",
      },
      {
        name: "data",
        type: "bytes32",
      },
    ],
    name: "fulfillOracleRequest",
    outputs: [
      {
        name: "",
        type: "bool",
      },
    ],
    payable: false,
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    constant: true,
    inputs: [],
    name: "withdrawable",
    outputs: [
      {
        name: "",
        type: "uint256",
      },
    ],
    payable: false,
    stateMutability: "view",
    type: "function",
  },
  {
    constant: false,
    inputs: [
      {
        name: "node",
        type: "address",
      },
      {
        name: "allowed",
        type: "bool",
      },
    ],
    name: "setFulfillmentPermission",
    outputs: [],
    payable: false,
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    constant: true,
    inputs: [
      {
        name: "node",
        type: "address",
      },
    ],
    name: "getAuthorizationStatus",
    outputs: [
      {
        name: "",
        type: "bool",
      },
    ],
    payable: false,
    stateMutability: "view",
    type: "function",
  },
  {
    constant: false,
    inputs: [
      {
        name: "recipient",
        type: "address",
      },
      {
        name: "amount",
        type: "uint256",
      },
    ],
    name: "withdraw",
    outputs: [],
    payable: false,
    stateMutability: "nonpayable",
    type: "function",
  },
];
