// @chainlink/contracts/ethers/v0.4/AggregatorV3InterfaceFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "ethers/providers";

import { AggregatorV3Interface } from "./AggregatorV3Interface";

export class AggregatorV3InterfaceFactory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): AggregatorV3Interface {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as AggregatorV3Interface;
  }
}

const _abi = [
  {
    constant: true,
    inputs: [],
    name: "decimals",
    outputs: [
      {
        name: "",
        type: "uint8",
      },
    ],
    payable: false,
    stateMutability: "view",
    type: "function",
  },
  {
    constant: true,
    inputs: [],
    name: "version",
    outputs: [
      {
        name: "",
        type: "uint256",
      },
    ],
    payable: false,
    stateMutability: "view",
    type: "function",
  },
  {
    constant: true,
    inputs: [],
    name: "description",
    outputs: [
      {
        name: "",
        type: "string",
      },
    ],
    payable: false,
    stateMutability: "view",
    type: "function",
  },
  {
    constant: true,
    inputs: [
      {
        name: "_roundId",
        type: "uint80",
      },
    ],
    name: "getRoundData",
    outputs: [
      {
        name: "roundId",
        type: "uint80",
      },
      {
        name: "answer",
        type: "int256",
      },
      {
        name: "startedAt",
        type: "uint256",
      },
      {
        name: "updatedAt",
        type: "uint256",
      },
      {
        name: "answeredInRound",
        type: "uint80",
      },
    ],
    payable: false,
    stateMutability: "view",
    type: "function",
  },
  {
    constant: true,
    inputs: [],
    name: "latestRoundData",
    outputs: [
      {
        name: "roundId",
        type: "uint80",
      },
      {
        name: "answer",
        type: "int256",
      },
      {
        name: "startedAt",
        type: "uint256",
      },
      {
        name: "updatedAt",
        type: "uint256",
      },
      {
        name: "answeredInRound",
        type: "uint80",
      },
    ],
    payable: false,
    stateMutability: "view",
    type: "function",
  },
];
