// @chainlink/contracts/ethers/v0.4/ChainlinkedFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, ContractFactory, Signer } from "ethers";
import { Provider } from "ethers/providers";
import { UnsignedTransaction } from "ethers/utils/transaction";

import { TransactionOverrides } from ".";
import { Chainlinked } from "./Chainlinked";

export class ChainlinkedFactory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: TransactionOverrides): Promise<Chainlinked> {
    return super.deploy(overrides) as Promise<Chainlinked>;
  }
  getDeployTransaction(overrides?: TransactionOverrides): UnsignedTransaction {
    return super.getDeployTransaction(overrides);
  }
  attach(address: string): Chainlinked {
    return super.attach(address) as Chainlinked;
  }
  connect(signer: Signer): ChainlinkedFactory {
    return super.connect(signer) as ChainlinkedFactory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): Chainlinked {
    return new Contract(address, _abi, signerOrProvider) as Chainlinked;
  }
}

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32",
      },
    ],
    name: "ChainlinkRequested",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32",
      },
    ],
    name: "ChainlinkFulfilled",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        name: "id",
        type: "bytes32",
      },
    ],
    name: "ChainlinkCancelled",
    type: "event",
  },
];

const _bytecode =
  "0x60806040526001600455348015601457600080fd5b5060358060226000396000f3006080604052600080fd00a165627a7a72305820d4a1f647c00ac7325f9a8a912598d4c4d0516a0ec523c5fb31ae354eb4ca1adc0029";
