// @0x/sol-resolver/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fallthrough_resolver_1 = require("./resolvers/fallthrough_resolver");
exports.FallthroughResolver = fallthrough_resolver_1.FallthroughResolver;
var url_resolver_1 = require("./resolvers/url_resolver");
exports.URLResolver = url_resolver_1.URLResolver;
var npm_resolver_1 = require("./resolvers/npm_resolver");
exports.NPMResolver = npm_resolver_1.NPMResolver;
var fs_resolver_1 = require("./resolvers/fs_resolver");
exports.FSResolver = fs_resolver_1.FSResolver;
var relative_fs_resolver_1 = require("./resolvers/relative_fs_resolver");
exports.RelativeFSResolver = relative_fs_resolver_1.RelativeFSResolver;
var name_resolver_1 = require("./resolvers/name_resolver");
exports.NameResolver = name_resolver_1.NameResolver;
var spy_resolver_1 = require("./resolvers/spy_resolver");
exports.SpyResolver = spy_resolver_1.SpyResolver;
var enumerable_resolver_1 = require("./resolvers/enumerable_resolver");
exports.EnumerableResolver = enumerable_resolver_1.EnumerableResolver;
var resolver_1 = require("./resolvers/resolver");
exports.Resolver = resolver_1.Resolver;
