// @0x/subproviders/lib/src/subproviders/private_key_wallet.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@0x/assert");
const utils_1 = require("@0x/utils");
const EthereumTx = require("ethereumjs-tx");
const ethUtil = require("ethereumjs-util");
const types_1 = require("../types");
const base_wallet_subprovider_1 = require("./base_wallet_subprovider");
/**
 * This class implements the [web3-provider-engine](https://github.com/MetaMask/provider-engine) subprovider interface.
 * This subprovider intercepts all account related RPC requests (e.g message/transaction signing, etc...) and handles
 * all requests with the supplied Ethereum private key.
 */
class PrivateKeyWalletSubprovider extends base_wallet_subprovider_1.BaseWalletSubprovider {
    /**
     * Instantiates a PrivateKeyWalletSubprovider.
     * @param privateKey The corresponding private key to an Ethereum address
     * @return PrivateKeyWalletSubprovider instance
     */
    constructor(privateKey) {
        assert_1.assert.isString('privateKey', privateKey);
        super();
        this._privateKeyBuffer = Buffer.from(privateKey, 'hex');
        this._address = `0x${ethUtil.privateToAddress(this._privateKeyBuffer).toString('hex')}`;
    }
    /**
     * Retrieve the account associated with the supplied private key.
     * This method is implicitly called when issuing a `eth_accounts` JSON RPC request
     * via your providerEngine instance.
     * @return An array of accounts
     */
    getAccountsAsync() {
        return __awaiter(this, void 0, void 0, function* () {
            return [this._address];
        });
    }
    /**
     * Sign a transaction with the private key. If you've added this Subprovider to your
     * app's provider, you can simply send an `eth_sendTransaction` JSON RPC request, and
     * this method will be called auto-magically. If you are not using this via a ProviderEngine
     * instance, you can call it directly.
     * @param txParams Parameters of the transaction to sign
     * @return Signed transaction hex string
     */
    signTransactionAsync(txParams) {
        return __awaiter(this, void 0, void 0, function* () {
            PrivateKeyWalletSubprovider._validateTxParams(txParams);
            if (txParams.from !== undefined && txParams.from.toLowerCase() !== this._address.toLowerCase()) {
                throw new Error(`Requested to sign transaction with address: ${txParams.from}, instantiated with address: ${this._address}`);
            }
            const tx = new EthereumTx(txParams);
            tx.sign(this._privateKeyBuffer);
            const rawTx = `0x${tx.serialize().toString('hex')}`;
            return rawTx;
        });
    }
    /**
     * Sign a personal Ethereum signed message. The signing address will be calculated from the private key.
     * The address must be provided it must match the address calculated from the private key.
     * If you've added this Subprovider to your app's provider, you can simply send an `eth_sign`
     * or `personal_sign` JSON RPC request, and this method will be called auto-magically.
     * If you are not using this via a ProviderEngine instance, you can call it directly.
     * @param data Hex string message to sign
     * @param address Address of the account to sign with
     * @return Signature hex string (order: rsv)
     */
    signPersonalMessageAsync(data, address) {
        return __awaiter(this, void 0, void 0, function* () {
            if (data === undefined) {
                throw new Error(types_1.WalletSubproviderErrors.DataMissingForSignPersonalMessage);
            }
            assert_1.assert.isHexString('data', data);
            assert_1.assert.isETHAddressHex('address', address);
            if (address.toLowerCase() !== this._address.toLowerCase()) {
                throw new Error(`Requested to sign message with address: ${address}, instantiated with address: ${this._address}`);
            }
            const dataBuff = ethUtil.toBuffer(data);
            const msgHashBuff = ethUtil.hashPersonalMessage(dataBuff);
            const sig = ethUtil.ecsign(msgHashBuff, this._privateKeyBuffer);
            const rpcSig = ethUtil.toRpcSig(sig.v, sig.r, sig.s);
            return rpcSig;
        });
    }
    /**
     * Sign an EIP712 Typed Data message. The signing address will be calculated from the private key.
     * The address must be provided it must match the address calculated from the private key.
     * If you've added this Subprovider to your app's provider, you can simply send an `eth_signTypedData`
     * JSON RPC request, and this method will be called auto-magically.
     * If you are not using this via a ProviderEngine instance, you can call it directly.
     * @param address Address of the account to sign with
     * @param data the typed data object
     * @return Signature hex string (order: rsv)
     */
    signTypedDataAsync(address, typedData) {
        return __awaiter(this, void 0, void 0, function* () {
            if (typedData === undefined) {
                throw new Error(types_1.WalletSubproviderErrors.DataMissingForSignTypedData);
            }
            assert_1.assert.isETHAddressHex('address', address);
            if (address.toLowerCase() !== this._address.toLowerCase()) {
                throw new Error(`Requested to sign message with address: ${address}, instantiated with address: ${this._address}`);
            }
            const dataBuff = utils_1.signTypedDataUtils.generateTypedDataHash(typedData);
            const sig = ethUtil.ecsign(dataBuff, this._privateKeyBuffer);
            const rpcSig = ethUtil.toRpcSig(sig.v, sig.r, sig.s);
            return rpcSig;
        });
    }
}
exports.PrivateKeyWalletSubprovider = PrivateKeyWalletSubprovider;
