// @nomiclabs/ethereumjs-vm/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const BN = require("bn.js");
const ethereumjs_account_1 = require("ethereumjs-account");
const ethereumjs_blockchain_1 = require("ethereumjs-blockchain");
const ethereumjs_common_1 = require("ethereumjs-common");
const state_1 = require("./state");
const runCode_1 = require("./runCode");
const runCall_1 = require("./runCall");
const runTx_1 = require("./runTx");
const runBlock_1 = require("./runBlock");
const opcodes_1 = require("./evm/opcodes");
const runBlockchain_1 = require("./runBlockchain");
const promisified_1 = require("./state/promisified");
const promisify = require('util.promisify');
const AsyncEventEmitter = require('async-eventemitter');
const Trie = require('merkle-patricia-tree/secure.js');
/**
 * Execution engine which can be used to run a blockchain, individual
 * blocks, individual transactions, or snippets of EVM bytecode.
 *
 * This class is an AsyncEventEmitter, please consult the README to learn how to use it.
 */
class VM extends AsyncEventEmitter {
    /**
     * Instantiates a new [[VM]] Object.
     * @param opts - Default values for the options are:
     *  - `chain`: 'mainnet'
     *  - `hardfork`: 'petersburg' [supported: 'byzantium', 'constantinople', 'petersburg', 'istanbul' (DRAFT) (will throw on unsupported)]
     *  - `activatePrecompiles`: false
     *  - `allowUnlimitedContractSize`: false [ONLY set to `true` during debugging]
     */
    constructor(opts = {}) {
        super();
        this.opts = opts;
        if (opts.common) {
            if (opts.chain || opts.hardfork) {
                throw new Error('You can only instantiate the VM class with one of: opts.common, or opts.chain and opts.hardfork');
            }
            this._common = opts.common;
        }
        else {
            const chain = opts.chain ? opts.chain : 'mainnet';
            const hardfork = opts.hardfork ? opts.hardfork : 'petersburg';
            const supportedHardforks = [
                'byzantium',
                'constantinople',
                'petersburg',
                'istanbul',
                'muirGlacier',
            ];
            this._common = new ethereumjs_common_1.default(chain, hardfork, supportedHardforks);
        }
        // Set list of opcodes based on HF
        this._opcodes = opcodes_1.getOpcodesForHF(this._common);
        if (opts.stateManager) {
            this.stateManager = opts.stateManager;
        }
        else {
            const trie = opts.state || new Trie();
            if (opts.activatePrecompiles) {
                for (let i = 1; i <= 8; i++) {
                    trie.put(new BN(i).toArrayLike(Buffer, 'be', 20), new ethereumjs_account_1.default().serialize());
                }
            }
            this.stateManager = new state_1.StateManager({ trie, common: this._common });
        }
        this.pStateManager = new promisified_1.default(this.stateManager);
        this.blockchain = opts.blockchain || new ethereumjs_blockchain_1.default({ common: this._common });
        this.allowUnlimitedContractSize =
            opts.allowUnlimitedContractSize === undefined ? false : opts.allowUnlimitedContractSize;
        // We cache this promisified function as it's called from the main execution loop, and
        // promisifying each time has a huge performance impact.
        this._emit = promisify(this.emit.bind(this));
    }
    /**
     * Processes blocks and adds them to the blockchain.
     *
     * This method modifies the state.
     *
     * @param blockchain -  A [blockchain](https://github.com/ethereum/ethereumjs-blockchain) object to process
     */
    runBlockchain(blockchain) {
        return runBlockchain_1.default.bind(this)(blockchain);
    }
    /**
     * Processes the `block` running all of the transactions it contains and updating the miner's account
     *
     * This method modifies the state. If `generate` is `true`, the state modifications will be
     * reverted if an exception is raised. If it's `false`, it won't revert if the block's header is
     * invalid. If an error is thrown from an event handler, the state may or may not be reverted.
     *
     * @param opts - Default values for options:
     *  - `generate`: false
     */
    runBlock(opts) {
        return runBlock_1.default.bind(this)(opts);
    }
    /**
     * Process a transaction. Run the vm. Transfers eth. Checks balances.
     *
     * This method modifies the state. If an error is thrown, the modifications are reverted, except
     * when the error is thrown from an event handler. In the latter case the state may or may not be
     * reverted.
     */
    runTx(opts) {
        return runTx_1.default.bind(this)(opts);
    }
    /**
     * runs a call (or create) operation.
     *
     * This method modifies the state.
     */
    runCall(opts) {
        return runCall_1.default.bind(this)(opts);
    }
    /**
     * Runs EVM code.
     *
     * This method modifies the state.
     */
    runCode(opts) {
        return runCode_1.default.bind(this)(opts);
    }
    /**
     * Returns a copy of the [[VM]] instance.
     */
    copy() {
        return new VM({
            stateManager: this.stateManager.copy(),
            blockchain: this.blockchain,
            common: this._common,
        });
    }
}
exports.default = VM;
